[Lim En Qi Angie] iP

LapisRaider opened the pull request:
# Duke but slightly passive-aggressive

> "Sixteen times the detail! IT JUST WORKS!" - [Todd Howard](https://en.wikipedia.org/wiki/Todd_Howard)

**Why you should use Duke:**
- text-based
- easy to learn
- ~~THE BETTER~~ THE VERY *GREATEST*!!

**Instructions to run Duke:**
1. Download it [here](https://github.com/LapisRaider/ip/releases/tag/A-Jar)
2. Open the file and run it
3. Start typing in the commands
4. IT JUST WORKS 👍 

**Features and commands:**
- [x] Add todo tasks `todo *insert task description here*`
- [x] Add events tasks `event *insert task description here* /at *date in yy-mm-dd format* *optional time in HHmm format*`
- [x] Add deadlines tasks `deadline *insert task description here* /by *date in yy-mm-dd format* *optional time in HHmm format*`
- [x] Lists out your tasklist `list`
- [x] Mark and unmark your tasklist `mark *task number*` or `unmark *task number*`
- [x] Delete task `delete *tasknumber*`
- [x] Save and load your tasklist (done automatically)

---
## Behind the scenes

Here's a little ***sneak peek*** of what runs Duke!!

```java
public static void main(String[] args) {
     Duke duke = new Duke();
     duke.runDuke();
}
```

Commit:
Add Gradle support

LapisRaider committed:
updated gitignore to ignore .class files

LapisRaider committed:
Finish Level 1. Able to Greet, Echo, Exit

LapisRaider committed:
Level-2, can add response into a list and print them out after typing command

LapisRaider committed:
Made function to print list to take in generics instead
Created the base for task class

LapisRaider committed:
- cleaned up Duke code to use member functions instead of static member functions
- seperated some of duke variable to become member variables

LapisRaider committed:
Level-3 is done, able to mark and unmark tasks now

LapisRaider committed:
added subclasses (Deadline, event and todo) for Task
make task member variables from private to protected so subclasses can inherit

LapisRaider committed:
remove the use of stringbuilder in task and its subclasses cause it defeats the purpose, normal string concat is about the same for small cases like this

LapisRaider committed:
fix some formatting issues when printing deadline and event
duke is able to handle todo, deadline and event commands
level-4 is done

LapisRaider committed:
remove the normal adding of task into list when there are no commands

LapisRaider committed:
just fix some formatting issues when printing for Duke

LapisRaider committed:
updated Expected.txt and input.txt for input testing
A-TextUiTesting

LapisRaider committed:
Place in exceptions for duke
- description of task cant be empty
- if forgot /by or /at for deadline and event
- if mark and unmark give an invlid task number
- if no valid command

LapisRaider committed:
- uipdated test cases to include the exception msg

LapisRaider committed:
Able to delete tasks, added exceptions for it and updated the test cases

LapisRaider committed:
ignore the actual.txt in test case

LapisRaider committed:
Add save functionality to tasks

Create Saving interface, made Task.java implement the interface

Add ability to save the tasklist base on modifications in Duke.java

LapisRaider committed:
Read tasklist from file on startup

Remove readFile() from Saving interface. Load should not be in save interface.

Add extra constructors in tasks and its child classes to be able to init isDone boolean.

read txt file on startup and initialise the values properly

LapisRaider committed:
Add date parsing

LapisRaider committed:
Add time parsing for event and deadline command

If user do not input a time, no warnings will be sent, it will set default timing as 2359 assuming to be due by or done at the end of the day.

LapisRaider committed:
Change filepath for storing tasklist data

LapisRaider committed:
Merge branch 'branch-Level-7'

LapisRaider committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

LapisRaider committed:
Extract tasklist to a seperated class

Extract tasklist and updated Duke's functions to use the new class. This makes the codes cleaner as everything related to tasklist such as modifying or editing its values is in a seperate file.

LapisRaider committed:
Add Storage class

Move loading and saving of tasklist from a file from Duke.java to Storage.java.

LapisRaider committed:
Extract file loading behavior to a seperate interface

This helps to make Storage.java more generic. Allows other class type to be able to use storage.java to store data, and also makes the class responsible for how it extracts data from the file for itself. In Duke.java made a lambda expression to pass to storage so it can create the proper task depending on what data received.

LapisRaider committed:
Remove unneeded imports in Duke.java

LapisRaider committed:
Extract Duke's Ui to its own class

LapisRaider committed:
Create the different commands classes

Start extracting out the commands from Duke.java to seperate command.
- Deadline Command
- Command
- EventCommand
- EditTaskMarkCommand
- ListCommand
- TaskCommand
- TodoCommand

LapisRaider committed:
Add byeCommand

LapisRaider committed:
Fix exception error bugs in some commands

Fix some UI printing for eventCommand and DeadlineCommand where they were printing out the wrong errors. Add byeCommand

LapisRaider committed:
Duke use a Parser to parse commands

Make the codes neater and extracts out the command implementation away from Duke and into a seperate class.

LapisRaider committed:
Remove original commandParse function in Duke

We are using the Parser class now to help parse commands, no longer a need for the original function in Duke.

LapisRaider committed:
Change break of runDuke to setting bool to false

LapisRaider committed:
Add data.txt to gitignore

LapisRaider committed:
Divide classes into different packages and update imports

LapisRaider committed:
Clean task class and it's child classes

Remove some unwanted constructors and do proper intialisation of time and date variables for deadline and event.

LapisRaider committed:
Unit test for tasks

Add unit test for deadline, event and todo

LapisRaider committed:
Add jar manifest and data txt file to gitignore

LapisRaider committed:
Add find task with keyword command

LapisRaider committed:
Add javadocs header for command classes

LapisRaider committed:
Javadocs for Tasks and DukeException

LapisRaider committed:
Javadocs for Duke, and utils.

LapisRaider committed:
Reformat files to match coding standard.

LapisRaider committed:
Merge branch 'branch-A-CodingStandard'

LapisRaider committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/command/EditTaskMarkCommand.java
#	src/main/java/duke/util/Loading.java
#	src/main/java/duke/util/Saving.java
#	src/main/java/duke/util/Storage.java

LapisRaider committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskList.java

brigittesantoso commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,54 @@
+        Task task = taskList.removeTask(taskIndex);
+        storage.saveList(taskList.getTaskList());
+        ui.printResponse(
+                "Got it, task has been removed: \n" + task.toString() + "\nNow you have " + taskList.getTaskListSize()
```
Line length still looks longer than 120 chars. Could possibly consider to break as such:
ui.printResponse("Got it, task has been removed: \n" 
                + task.toString() + "\nNow you have " + taskList.getTaskListSize()
                                + " tasks in your list.");

LapisRaider replied:
Thank you for letting me know! I'll adjust it!

brigittesantoso commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,26 @@
+
+        TaskList filteredTasks = taskList.getTaskListWithKeyword(taskDescription);
+        if (filteredTasks.getTaskListSize() == 0) {
+            ui.printResponse("No tasks with the keyword : " + taskDescription + "\nYou need to put the exact keyword, it is not case sensitive.\n");
```
Line length longer than 120 chars, consider placing line break to improve readability.

jiewei98 replied:
Agree with what brigittesantoso said, could consider adding a line break to make the code more readable

LapisRaider replied:
Thank you for letting me know! I'll adjust it!

brigittesantoso commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,90 @@
+    public String saveFileFormat() {
+        return DEADLINE_SYMBOL + "|" + this.isDone + "|" + taskDescription + "|" + this.deadlineDate.format(
+                DateTimeFormatter.ofPattern("MMM d yyyy")) + "|" + this.deadlineTime.format(
+                DateTimeFormatter.ofPattern("hh:mm a")) + "\n";
```
Indentation for wrapped lines should be 8 spaces (i.e. twice the normal indentation of 4 spaces) more than the parent line. Line 56 should be 12 spaces instead of 8 spaces as it is wrapped inside. Similar issue in Event.java file.
![image](https://user-images.githubusercontent.com/80502407/151954982-fa409e76-cef3-4f38-9f6e-2d6e44662e58.png)

LapisRaider replied:
Thank you for letting me know! I'll adjust it!

brigittesantoso commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    public <T extends Loading> void loadFromSave(ArrayList<T> list,
+            Function<String, T> factory)
+            throws DukeException {
```
This line can just be combined with the previous line. I don't think there is a need for line break here.

LapisRaider replied:
Thanks for letting me know! I thought it was over 120 characters but it was lesser than that XD

brigittesantoso commented on src/main/java/duke/util/Saving.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Interface for saving data by making it into a string to be saved into a file.
+ */
+public interface Saving {
```
This interface file might not be necessary as you can just do it in Storage class.

LapisRaider replied:
Thanks for the suggestion, however, I was going for a more generic OOP approach. I wanted Storage class to be as generic as possible and be able to save and load other classes besides Task (although I know for this individual project we don't need to do that).

Thus, I created a Saving and Loading interface so the class that implements them can have their own behaviour or formatting/reading for saving and loading. I separated Saving and Loading into different interfaces as I felt it would be better too.

Thanks for the suggestion!

brigittesantoso commented on src/main/java/duke/util/Loading.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Interface for loading and extracting data from a string.
+ */
+public interface Loading {
```
This interface file might not be necessary as you can just do it in Storage class.

brigittesantoso commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,67 @@
+ * <p>e.g., visit new theme park.</>
+ */
+public class Todo extends Task {
+    static final char TODO_SYMBOL = 'T';
```
Class variable scope not declared and should be declared to non-public.
![image](https://user-images.githubusercontent.com/80502407/151956119-7094f1c6-d0f0-4a88-903a-7fdcac076dec.png)

LapisRaider replied:
Thank you! I missed that.

brigittesantoso commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,91 @@
+import java.time.format.DateTimeParseException;
+
+/**
+ * Add event command.
```
For header comments, the first sentence should start in the form Returns ..., Sends ..., Adds ... (not Add or Adding etc.)

LapisRaider replied:
Thank you for letting me know! I'll adjust it!

brigittesantoso commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,114 @@
+        try {
+            // write data to file
+            FileWriter fileWriter = new FileWriter(file);
+            for (Saving item : list) {
```
Spacing between item and : is unnecessary.

LapisRaider replied:
Sorry, I know the code will still compile and run without the space between item and :, but may I know if without space is recommended by the style guide? I can't find any info about it on the java coding standard page and most of the documentation I search through showed a space in between to improve readability.

brigittesantoso reviewed (commented):
Good job! Just some small minor changes needed 👍

jiewei98 commented on src/main/java/duke/command/EditTaskMarkCommand.java:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Command to edit the task to be either marked or unmarked.
+ */
+public class EditTaskMarkCommand extends Command {
```
Can possibly consider changing the class name into MarkTaskCommand to shorten it?

LapisRaider replied:
Thanks for the suggestion! I named it EditTaskMarkCommand instead as it is able to mark and unmark tasks. I was worried that if I just put MarkTaskCommand, people would assume it can only mark task and not unmark it 😅

jiewei98 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,90 @@
+     *             Data format: D|true/false|taskDescription|MMM d yyyy|hh:mm a
+     */
+    @Override
+    public void extractFileData(String data) {
```
extractFileData seems to suggest that we are extracting data from a file when in fact it is just a string so perhaps renaming it into extractDeadlineDetails would be more applicable?

LapisRaider replied:
Thanks for the suggestion! I'll change it to extractDataFromLine instead.

jiewei98 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,91 @@
+     *             Data format: E|true/false|taskDescription|MMM d yyyy|hh:mm a
+     */
+    @Override
+    public void extractFileData(String data) {
```
Same concern as Deadline, could maybe consider renaming it to extractDeadlineDetails?

jiewei98 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+ */
+public class TaskList {
+    // tasklist exception messages
+    private final static String INDEX_NO_EXIST = "Yo yo yo, this task don't exist. Give a legitimate task number.";
```
Could be changed to INVALID_INDEX_TEXT instead?

LapisRaider replied:
Thanks for the suggestion! I'll change it!

jiewei98 reviewed (commented):
LGTM. Few nits to fix as pointed out by brigittesantoso.

LapisRaider committed:
Reformat files to match coding standard.

LapisRaider committed:
Merge branch 'add-gradle-support'

LapisRaider committed:
Update test files to use updated function name.

LapisRaider committed:
Build with gradle.

LapisRaider committed:
Add checkstyle support.

LapisRaider committed:
Set up JavaFX.

LapisRaider committed:
Add basic JavaFX layout.

LapisRaider committed:
Add dialoguebox and update Duke GUI.

LapisRaider committed:
Modify code to use FXML instead.

LapisRaider committed:
Modify Duke to return its resopnses with FXML.

Change command classes to return a string and abstracted the UI printing from command class functions to Duke instead. This makes it so Duke is able to work for both JavaFX and normal text output printing.

LapisRaider committed:
Beautify UI.

Add new background, new profile images, and make the image to a circle instead. Text should not have border and background.

LapisRaider committed:
Improve UI with small adjustments.

LapisRaider committed:
Fix issue with dialogue box shrinking.

LapisRaider committed:
Use JavaFX 11 instead.

This help to remove the warnings since the current project uses JavaFX 11.

LapisRaider committed:
Modify the window.

Add title and icon to the window. Also make it so the window cannot be resized.

LapisRaider committed:
Clean up the codes for GUI.

Move default textUI file to UI folder. Change private to private final.

LapisRaider committed:
Merge branch 'branch-Level-10'

LapisRaider committed:
Refactor event and deadline class.

Various magic values in event and deadline class. Can cause confusion on what the values mean and make it harder to debug issues in the future.

Refactor them to no longer use magic numbers so it'll be clearer to understand and easier to change the values in the future. This will improve code quality.

Let's abstract out the magic values and declare them as named constants.

LapisRaider committed:
Remove complicated expressions.

Various long and complicated expressions in various command classes.

Long complicated expressions, makes it hard to read, understand and debug the code. It is also prone to bugs if the code were to get longer.

Let's abstract and seperat out the complicated expressions to simpler more readble expressions to improve code quality.

LapisRaider committed:
Refactor codes to make code quality better.

LapisRaider committed:
Add assertions to command class.

Command class getDescription function able to take in an empty input, or input without the proper command key.

Will cause a parsing bug and obtain the wrong data as the parsing of data requires heavily on the key being inside the input, or for the input to not be empty.

Let's add assertions to command class getDescription function, to ensure the key exists in the input passed in, and that the input is not empty.

LapisRaider committed:
Merge pull request #2 from LapisRaider/branch-A-Assertions

Add assertions.

LapisRaider committed:
Merge pull request #1 from LapisRaider/branch-A-CodeQuality

Refactor codes to improve code quality.

LapisRaider committed:
Add github actions workflow.

LapisRaider committed:
Add gradlew and modify gradle.

Github Actions currently not working when it comes to building due to missing gradlew file. Current gradle file mainclassname is also not updated to the latest.

Need to add back the gradlew file and modify the gradle file so github action workflow will be able to run and this will make the work process much easier.

Let's add back the gradlew file.

LapisRaider committed:
Update github workflow action file.

Change wrapper permission for gradlew.

LapisRaider committed:
Modify line 33 of gihub action file.

LapisRaider committed:
Add more flexibility in search.

Users have to input full keywords to be able to find the tasks they want.

This can be irritating if users want to find items based on inputting the keyword partially. Being able to input the keyword partially to find the items they want will also provide a better user experience.

Let's change the find task function to accept partial keywords instead of full keywords.

LapisRaider committed:
Fix all checkstyle issues.

LapisRaider committed:
Fix issue with filepath.

LapisRaider committed:
Add Manifest file for generating jar.

LapisRaider committed:
Remove IO redirection in github workflow.

LapisRaider committed:
Add different response when user types wrong command.

LapisRaider committed:
Refactor code to constants instead of magic variables.

LapisRaider committed:
Close application with input "bye".

LapisRaider committed:
Refactor Main class to use constants.

LapisRaider committed:
Fix checkstyle issues.

LapisRaider committed:
Add and clean method header comments.

LapisRaider committed:
Change protected to private.

LapisRaider committed:
Add representative image.

LapisRaider committed:
Set theme jekyll-theme-architect

LapisRaider committed:
Create help command and place keys in their command class.

LapisRaider committed:
Fix checkstyle errors.

LapisRaider committed:
Add javadocs for help command.

LapisRaider committed:
Edit README to show features.

LapisRaider committed:
Empty tasklist if there's error loading file.

LapisRaider committed:
Write user guide.

LapisRaider committed:
Fix issue with loading in Linux.

LapisRaider committed:
Add header comment to storage

LapisRaider committed:
Update userguide

LapisRaider committed:
Fix minor wording issues

LapisRaider committed:
Fix header comment issues

LapisRaider committed:
Minor spelling change in userguide
